### DISCUSSION TITLE
Partially paired randomization of img and video?

### DISCUSSION THREAD

**Discussion by Kedersha** [jspsych/jsPsych#1211] [Q&A] [answered]
Hello! I'm developing an experiment and am finding the randomization options to be slightly confusing; I'm not sure which would be the best for my use case, and was hoping to get some advice.

Design of study:

- **Learning block**: video-keyboard-response, participants see short video randomly paired with an image (they are learning to associate a pseudo-sign from the video with the object in the picture)
- **Experimental block**: video-keyboard-response + html-button-response, participants see video then choose between two options of images

Goals with randomization:

- At start of study, **randomly associate each video with 1 target image** and maintain that association through the study (this is the part I'm really struggling with!)
- **Shuffle order of videos** between learning block & experimental block
- During experimental block, **randomly select distractor image** (sample without replacement, not identical to the target image)

Right now, this is how the code for the learning and experimental trials looks (with abbreviated stimulus lists for space); it runs, but doesn't meet any of my randomization goals:

```js
      //Randomize Video
      var e_video_list = [['shared_assets_nsl/video/silent0.mp4'],
      		['shared_assets/video/silent1.mp4'],
      		['shared_assets/video/silent2.mp4']];
      var e_video = jsPsych.randomization.shuffle(e_video_list);
      
	var e_img_list = [['shared_assets_nsl/img/0.jpg','shared_assets/img/1.png'],
		['shared_assets/img/1.jpg','shared_assets/img/2.png'],
		['shared_assets/img/2.jpg','shared_assets/img/3.png']];
		var e_img = jsPsych.randomization.shuffle(e_img_list);

//LEARNING BLOCK
	for (let i = 0; i <= 2; i++){
		//Play Video
		let t_trial = {
			type: 'video-keyboard-response',
			sources: e_video[i],
			prompt: '<img src=' + e_img[i][0] + ' width="200px" height="200px">',
			choices: jsPsych.NO_KEYS,
			width: 640,
			trial_ends_after_video: true,
			data: {e_target: e_video[i][0].replace(/shared_assets\/video\//gi, "")}
		};
		timeline.push(t_trial);
	};

//EXPERIMENTAL BLOCK
      for (let i = 0; i <= 2; i++){
        //Play Video
        let e_trial = {
            type: 'video-keyboard-response',
            sources: e_video[i],
	choices: jsPsych.NO_KEYS,
            width: 640,
            trial_ends_after_video: true,
            data: {e_target: e_video[i][0].replace(/shared_assets\/video\//gi, "")}
          };

        //Choose image
        let e_response = {
            type: 'html-button-response',
            stimulus: ' ',
            prompt:'Choose an image',
            choices: e_img[i],
            button_html: '<button class="jspsych-btn"><img src=%choice% width="100px" height="100px"></button>',
            margin_vertical: '100px',
            margin_horizontal: '100px',
		data: {e_target: e_img[i].toString().replace(/shared_assets_nsl\/img\//gi, "")}          
	};
        timeline.push(e_trial, e_response);
};
```

I've started trying to implement a solution using partially-constructed trials and jsPsych.randomization.repeat(trial_list,1), but it's not coming together well--it doesn't randomize that initial video/img association, and even so, I'm still unable to fetch the information that I actually want from the randomized list:

```js
var trial1 = {
	sources: ['shared_assets/video/silent0.mp4'],
	choices: ['shared_assets/img/0.png'],
	target: '0'
};
var trial2 = {
	sources: ['shared_assets/video/silent1.mp4'],
	choices: ['shared_assets/img/1.png'],
	target: '1'
};
var trial3 = {
	sources: ['shared_assets/video/silent2.mp4'],
	choices: ['shared_assets/img/2.png'],
	target: '2'
};
var t_trial_list = {trial1, trial2, trial3}
var t_trial = jsPsych.randomization.repeat(t_trial_list,1);
```

If anyone has advice, I'd really appreciate it! Thanks!

**Comment by jodeleeuw** [comment, marked as answer]
Hi @Kedersha ,

I think this is a situation that would really benefit from [timeline variables](https://www.jspsych.org/overview/timeline/#timeline-variables).

You can use code similar to what you have now to build a list of timeline variables:

```js
// run the code to randomly shuffle images and videos first

// then...

var TOTAL_PAIRS = 4;
var img_video_pairs = [];
for(var i=0; i<TOTAL_PAIRS; i++){
	img_video_pairs.push({
		video: e_video[i],
		img_target: e_img[i][1],
		img_distractor: e_img[i][2]
	});
}
```

This should create a stable set of pairs that doesn't change throughout the experiment.

Then in the two blocks, you can use `randomize_order: true` in the timeline with the timeline variables. For example, here's what the learning block could look like:

```js
var learning_block = {
	timeline: [{
		type: 'video-keyboard-response',
		sources: jsPsych.timelineVariable('video'),
		prompt: function(){ return '<img src=' + jsPsych.timelineVariable('target', true) + ' width="200px" height="200px">' },
		choices: jsPsych.NO_KEYS,
		width: 640,
		trial_ends_after_video: true,
		data: {
			e_target: function(){ return jsPsych.timelineVariable('video', true)[0].replace(/shared_assets\/video\//gi, "") }
		}
	}],
	timeline_variables: img_video_pairs,
	randomize_order: true
}
```

A similar approach should work for the test phase. 

> During experimental block, randomly select distractor image (sample without replacement, not identical to the target image)

If each video-img pairing has a unique distractor image, then you could add the distractor to the timeline variables like I've done above. Let us know if you need to do something different and we can provide further guidance.

**Comment by Kedersha** [comment]
Thanks for your help, @jodeleeuw ! Your suggestions and code snippets have helped me out a lot--timeline variables do indeed make this a lot more straightforward. If you have time/inclination to help steer this a little further, I'd be grateful, but if not no worries, I can also keep plugging away.

My current steps are:

- make it so the distractor image changes from one run of the study to the next; we don't want to use the same distractor image for a given target image for every participant [**ETA: this is working, I'll leave my code at the bottom of this message in case anyone's reading this thread with a similar problem**]
- for experimental trials, randomize from trial to trial whether the target image is on the left or on the right [**ETA: also working! I should've specified my randomized choices as part of the timeline variable**]
 - finally, write to trial data what the jpg was for the participant's choice/whether the participant chose correctly [**ETA: got this working, too!**]

For the record, here's the working version of the relevant pieces:

```
var e_img_list = ['shared_assets/img/1.jpg','shared_assets/img/2.jpg','shared_assets/img/3.jpg','shared_assets/img/4.jpg'];

var e_img_target = jsPsych.randomization.shuffle(e_img_list);
var e_img_distract = jsPsych.randomization.shuffle(e_img_list);

//check for cases where target and distractor match; re-shuffle if that's the case
var dupes = false;
for(var j=0; j<e_img_target.length; j++){
	if(e_img_target[j] == e_img_distract[j] || e_img_target[0] == e_img_distract[0]){
		dupes = true;
	}
	if(dupes){
		var e_img_distract = jsPsych.randomization.shuffle(e_img_list);
		j=0;
		dupes = false;
	}
}

var TOTAL_TRIOS = 4;
var img_video_trios = [];
for(var i=0; i<TOTAL_TRIOS; i++){
	choicearray = jsPsych.randomization.shuffle([e_img_target[i],e_img_distract[i]]);
	targetindex = choicearray.indexOf(e_img_target[i]);
	img_video_trios.push({
		video: e_video[i],
		img_target: e_img_target[i],
		img_distractor: e_img_distract[i]
		choices: choicearray,
		data: {
			target_index: targetindex,
			target_image: e_img_target[i],
			distractor_image: e_img_distract[i],
			video: e_video[i]
			}
		});
};

```

[timeline management, learning blocks, experiment instruction...]

```
//Function to check whether the response was accurate or not
function check_correct(data) {
	var correct = false;
	if (data.button_pressed == data.target_index) {
		correct = true;
	}
	data.correct = correct;
};

//EXPERIMENTAL BLOCK: image-button-response
var exp_block = {
	timeline: [{//stimulus presentation
		type: 'video-keyboard-response',
		sources: jsPsych.timelineVariable('video'),
		choices: jsPsych.NO_KEYS,
		width: 500,
		trial_ends_after_video: true,
		data: {
			video: function(){ return jsPsych.timelineVariable('video', true)[0]}
			}
		},
		{//response
		type: "image-button-response",
		stimulus: '',
		choices: jsPsych.timelineVariable('choices'),
		button_html: '<button class="jspsych-btn"><img src=%choice% width="200px" height="200px"></button>',
		prompt: "<p>Click the image was associated with the sign.</p>",
		data: jsPsych.timelineVariable('data'),
		on_finish: check_correct
		}
	],
	timeline_variables: img_video_trios,
	randomize_order: true
};
timeline.push(exp_block);
```
